File: server/internal/services/ai_service.go
package services

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strings"
	"time"

	"notesage-server/internal/models"

	"github.com/google/uuid"
	"gorm.io/gorm"
)

// AIProvider represents different AI service providers
type AIProvider string

const (
	ProviderOpenAI AIProvider = "openai"
	ProviderGemini AIProvider = "gemini"
	ProviderGrok   AIProvider = "grok"
)

// AIConfig holds configuration for AI providers
type AIConfig struct {
	Provider   AIProvider `json:"provider"`
	APIKey     string     `json:"api_key"`
	BaseURL    string     `json:"base_url,omitempty"`
	Model      string     `json:"model,omitempty"`
	MaxTokens  int        `json:"max_tokens,omitempty"`
	Timeout    int        `json:"timeout,omitempty"` // seconds
}

// AIService handles AI-powered features
type AIService struct {
	db       *gorm.DB
	config   *AIConfig
	client   *http.Client
	enabled  bool
}

// NewAIService creates a new AI service
func NewAIService(db *gorm.DB, config *AIConfig) *AIService {
	timeout := 30 * time.Second
	if config != nil && config.Timeout > 0 {
		timeout = time.Duration(config.Timeout) * time.Second
	}

	return &AIService{
		db:      db,
		config:  config,
		enabled: config != nil && config.APIKey != "",
		client: &http.Client{
			Timeout: timeout,
		},
	}
}

// IsEnabled returns whether AI features are available
func (s *AIService) IsEnabled() bool {
	return s.enabled
}

// TodoExtractionResult represents the result of AI todo extraction
type TodoExtractionResult struct {
	Todos  []ExtractedTodo `json:"todos"`
	Error  string          `json:"error,omitempty"`
}

// ExtractedTodo represents a todo extracted by AI
type ExtractedTodo struct {
	Text             string `json:"text"`
	AssignedPersonID string `json:"assigned_person_id,omitempty"`
	DueDate          string `json:"due_date,omitempty"` // YYYY-MM-DD format
	Priority         string `json:"priority,omitempty"`
	Context          string `json:"context,omitempty"`
}

// PeopleAnalysisResult represents the result of AI people analysis
type PeopleAnalysisResult struct {
	Mentions     []AIPersonMention   `json:"mentions"`
	Relationships []PersonRelationship `json:"relationships"`
	Error        string              `json:"error,omitempty"`
}

// AIPersonMention represents a person mentioned in content (AI analysis)
type AIPersonMention struct {
	Name     string `json:"name"`
	Context  string `json:"context"`
	Strength int    `json:"strength"` // 1-10 scale
}

// PersonRelationship represents a relationship between people
type PersonRelationship struct {
	Person1      string `json:"person1"`
	Person2      string `json:"person2"`
	Relationship string `json:"relationship"`
	Confidence   float64 `json:"confidence"`
}

// InsightResult represents AI-generated insights
type InsightResult struct {
	Insights []Insight `json:"insights"`
	Error    string    `json:"error,omitempty"`
}

// Insight represents a knowledge pattern insight
type Insight struct {
	Type        string  `json:"type"`        // "pattern", "connection", "trend", "suggestion"
	Title       string  `json:"title"`
	Description string  `json:"description"`
	Confidence  float64 `json:"confidence"`
	Data        map[string]interface{} `json:"data,omitempty"`
}

// ExtractTodosFromNote extracts todos from note content using AI
func (s *AIService) ExtractTodosFromNote(ctx context.Context, noteContent models.JSONB) (*TodoExtractionResult, error) {
	if !s.enabled {
		return &TodoExtractionResult{
			Error: "AI service not available",
		}, nil
	}

	// Extract text content from structured JSON
	textContent := s.extractTextFromContent(noteContent)
	if textContent == "" {
		return &TodoExtractionResult{Todos: []ExtractedTodo{}}, nil
	}

	prompt := s.buildTodoExtractionPrompt(textContent)
	response, err := s.callAI(ctx, prompt)
	if err != nil {
		return &TodoExtractionResult{
			Error: fmt.Sprintf("AI request failed: %v", err),
		}, nil
	}

	return s.parseTodoExtractionResponse(response)
}

// AnalyzePeopleMentions analyzes people mentions and relationships in content
func (s *AIService) AnalyzePeopleMentions(ctx context.Context, noteContent models.JSONB, userID uuid.UUID) (*PeopleAnalysisResult, error) {
	if !s.enabled {
		return &PeopleAnalysisResult{
			Error: "AI service not available",
		}, nil
	}

	textContent := s.extractTextFromContent(noteContent)
	if textContent == "" {
		return &PeopleAnalysisResult{
			Mentions:     []AIPersonMention{},
			Relationships: []PersonRelationship{},
		}, nil
	}

	// Get existing people for context
	var existingPeople []models.Person
	s.db.Where("user_id = ?", userID).Find(&existingPeople)

	prompt := s.buildPeopleAnalysisPrompt(textContent, existingPeople)
	response, err := s.callAI(ctx, prompt)
	if err != nil {
		return &PeopleAnalysisResult{
			Error: fmt.Sprintf("AI request failed: %v", err),
		}, nil
	}

	return s.parsePeopleAnalysisResponse(response)
}

// GenerateInsights generates insights from user's knowledge base
func (s *AIService) GenerateInsights(ctx context.Context, userID uuid.UUID, limit int) (*InsightResult, error) {
	if !s.enabled {
		return &InsightResult{
			Error: "AI service not available",
		}, nil
	}

	// Gather user's data for analysis
	userData, err := s.gatherUserData(userID, limit)
	if err != nil {
		return &InsightResult{
			Error: fmt.Sprintf("Failed to gather user data: %v", err),
		}, nil
	}

	prompt := s.buildInsightGenerationPrompt(userData)
	response, err := s.callAI(ctx, prompt)
	if err != nil {
		return &InsightResult{
			Error: fmt.Sprintf("AI request failed: %v", err),
		}, nil
	}

	return s.parseInsightResponse(response)
}

// callAI makes a request to the configured AI provider
func (s *AIService) callAI(ctx context.Context, prompt string) (string, error) {
	switch s.config.Provider {
	case ProviderOpenAI:
		return s.callOpenAI(ctx, prompt)
	case ProviderGemini:
		return s.callGemini(ctx, prompt)
	case ProviderGrok:
		return s.callGrok(ctx, prompt)
	default:
		return "", fmt.Errorf("unsupported AI provider: %s", s.config.Provider)
	}
}

// callOpenAI makes a request to OpenAI API
func (s *AIService) callOpenAI(ctx context.Context, prompt string) (string, error) {
	baseURL := s.config.BaseURL
	if baseURL == "" {
		baseURL = "https://api.openai.com/v1"
	}

	model := s.config.Model
	if model == "" {
		model = "gpt-3.5-turbo"
	}

	maxTokens := s.config.MaxTokens
	if maxTokens == 0 {
		maxTokens = 1000
	}

	requestBody := map[string]interface{}{
		"model": model,
		"messages": []map[string]string{
			{
				"role":    "user",
				"content": prompt,
			},
		},
		"max_tokens":  maxTokens,
		"temperature": 0.7,
	}

	jsonBody, err := json.Marshal(requestBody)
	if err != nil {
		return "", fmt.Errorf("failed to marshal request: %w", err)
	}

	req, err := http.NewRequestWithContext(ctx, "POST", baseURL+"/chat/completions", bytes.NewBuffer(jsonBody))
	if err != nil {
		return "", fmt.Errorf("failed to create request: %w", err)
	}

	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Authorization", "Bearer "+s.config.APIKey)

	resp, err := s.client.Do(req)
	if err != nil {
		return "", fmt.Errorf("request failed: %w", err)
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return "", fmt.Errorf("failed to read response: %w", err)
	}

	if resp.StatusCode != http.StatusOK {
		return "", fmt.Errorf("API request failed with status %d: %s", resp.StatusCode, string(body))
	}

	var response struct {
		Choices []struct {
			Message struct {
				Content string `json:"content"`
			} `json:"message"`
		} `json:"choices"`
	}

	if err := json.Unmarshal(body, &response); err != nil {
		return "", fmt.Errorf("failed to parse response: %w", err)
	}

	if len(response.Choices) == 0 {
		return "", fmt.Errorf("no response from AI")
	}

	return response.Choices[0].Message.Content, nil
}

// callGemini makes a request to Google Gemini API
func (s *AIService) callGemini(ctx context.Context, prompt string) (string, error) {
	baseURL := s.config.BaseURL
	if baseURL == "" {
		baseURL = "https://generativelanguage.googleapis.com/v1beta"
	}

	model := s.config.Model
	if model == "" {
		model = "gemini-pro"
	}

	requestBody := map[string]interface{}{
		"contents": []map[string]interface{}{
			{
				"parts": []map[string]string{
					{
						"text": prompt,
					},
				},
			},
		},
	}

	jsonBody, err := json.Marshal(requestBody)
	if err != nil {
		return "", fmt.Errorf("failed to marshal request: %w", err)
	}

	url := fmt.Sprintf("%s/models/%s:generateContent?key=%s", baseURL, model, s.config.APIKey)
	req, err := http.NewRequestWithContext(ctx, "POST", url, bytes.NewBuffer(jsonBody))
	if err != nil {
		return "", fmt.Errorf("failed to create request: %w", err)
	}

	req.Header.Set("Content-Type", "application/json")

	resp, err := s.client.Do(req)
	if err != nil {
		return "", fmt.Errorf("request failed: %w", err)
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return "", fmt.Errorf("failed to read response: %w", err)
	}

	if resp.StatusCode != http.StatusOK {
		return "", fmt.Errorf("API request failed with status %d: %s", resp.StatusCode, string(body))
	}

	var response struct {
		Candidates []struct {
			Content struct {
				Parts []struct {
					Text string `json:"text"`
				} `json:"parts"`
			} `json:"content"`
		} `json:"candidates"`
	}

	if err := json.Unmarshal(body, &response); err != nil {
		return "", fmt.Errorf("failed to parse response: %w", err)
	}

	if len(response.Candidates) == 0 || len(response.Candidates[0].Content.Parts) == 0 {
		return "", fmt.Errorf("no response from AI")
	}

	return response.Candidates[0].Content.Parts[0].Text, nil
}

// callGrok makes a request to Grok API (X.AI)
func (s *AIService) callGrok(ctx context.Context, prompt string) (string, error) {
	baseURL := s.config.BaseURL
	if baseURL == "" {
		baseURL = "https://api.x.ai/v1"
	}

	model := s.config.Model
	if model == "" {
		model = "grok-beta"
	}

	maxTokens := s.config.MaxTokens
	if maxTokens == 0 {
		maxTokens = 1000
	}

	requestBody := map[string]interface{}{
		"model": model,
		"messages": []map[string]string{
			{
				"role":    "user",
				"content": prompt,
			},
		},
		"max_tokens":  maxTokens,
		"temperature": 0.7,
	}

	jsonBody, err := json.Marshal(requestBody)
	if err != nil {
		return "", fmt.Errorf("failed to marshal request: %w", err)
	}

	req, err := http.NewRequestWithContext(ctx, "POST", baseURL+"/chat/completions", bytes.NewBuffer(jsonBody))
	if err != nil {
		return "", fmt.Errorf("failed to create request: %w", err)
	}

	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Authorization", "Bearer "+s.config.APIKey)

	resp, err := s.client.Do(req)
	if err != nil {
		return "", fmt.Errorf("request failed: %w", err)
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return "", fmt.Errorf("failed to read response: %w", err)
	}

	if resp.StatusCode != http.StatusOK {
		return "", fmt.Errorf("API request failed with status %d: %s", resp.StatusCode, string(body))
	}

	var response struct {
		Choices []struct {
			Message struct {
				Content string `json:"content"`
			} `json:"message"`
		} `json:"choices"`
	}

	if err := json.Unmarshal(body, &response); err != nil {
		return "", fmt.Errorf("failed to parse response: %w", err)
	}

	if len(response.Choices) == 0 {
		return "", fmt.Errorf("no response from AI")
	}

	return response.Choices[0].Message.Content, nil
}

// Helper methods for building prompts and parsing responses

func (s *AIService) buildTodoExtractionPrompt(content string) string {
	return fmt.Sprintf(`
Analyze the following note content and extract actionable todo items. Return a JSON response with the following structure:

{
  "todos": [
    {
      "text": "Complete the task description",
      "assigned_person_id": "person_name_if_mentioned",
      "due_date": "YYYY-MM-DD_if_date_mentioned",
      "priority": "high|medium|low",
      "context": "brief_context_from_note"
    }
  ]
}

Rules:
1. Only extract clear, actionable tasks
2. Include person names if someone is assigned or mentioned in relation to the task
3. Extract dates in YYYY-MM-DD format if mentioned
4. Assign priority based on urgency indicators in the text
5. Provide brief context about where the todo was found

Note content:
%s

Return only valid JSON, no additional text.`, content)
}

func (s *AIService) buildPeopleAnalysisPrompt(content string, existingPeople []models.Person) string {
	peopleNames := make([]string, len(existingPeople))
	for i, person := range existingPeople {
		peopleNames[i] = person.Name
	}

	return fmt.Sprintf(`
Analyze the following note content for people mentions and relationships. Return a JSON response with the following structure:

{
  "mentions": [
    {
      "name": "Person Name",
      "context": "context_where_mentioned",
      "strength": 5
    }
  ],
  "relationships": [
    {
      "person1": "Person A",
      "person2": "Person B", 
      "relationship": "colleague|friend|manager|client|etc",
      "confidence": 0.8
    }
  ]
}

Known people in the system: %s

Rules:
1. Identify all person mentions (names, pronouns referring to people)
2. Rate mention strength 1-10 based on importance in context
3. Identify relationships between people mentioned
4. Use confidence 0.0-1.0 for relationship certainty
5. Match against known people when possible

Note content:
%s

Return only valid JSON, no additional text.`, strings.Join(peopleNames, ", "), content)
}

func (s *AIService) buildInsightGenerationPrompt(userData map[string]interface{}) string {
	userDataJSON, _ := json.Marshal(userData)
	
	return fmt.Sprintf(`
Analyze the following user knowledge base data and generate insights. Return a JSON response with the following structure:

{
  "insights": [
    {
      "type": "pattern|connection|trend|suggestion",
      "title": "Insight Title",
      "description": "Detailed description of the insight",
      "confidence": 0.8,
      "data": {
        "additional": "context_data"
      }
    }
  ]
}

Rules:
1. Identify patterns in note-taking behavior
2. Find interesting connections between people and topics
3. Spot trends over time
4. Suggest improvements or actions
5. Rate confidence 0.0-1.0 for each insight
6. Limit to most valuable insights (max 5)

User data:
%s

Return only valid JSON, no additional text.`, string(userDataJSON))
}

func (s *AIService) parseTodoExtractionResponse(response string) (*TodoExtractionResult, error) {
	var result TodoExtractionResult
	if err := json.Unmarshal([]byte(response), &result); err != nil {
		return &TodoExtractionResult{
			Error: fmt.Sprintf("Failed to parse AI response: %v", err),
		}, nil
	}
	return &result, nil
}

func (s *AIService) parsePeopleAnalysisResponse(response string) (*PeopleAnalysisResult, error) {
	var result PeopleAnalysisResult
	if err := json.Unmarshal([]byte(response), &result); err != nil {
		return &PeopleAnalysisResult{
			Error: fmt.Sprintf("Failed to parse AI response: %v", err),
		}, nil
	}
	return &result, nil
}

func (s *AIService) parseInsightResponse(response string) (*InsightResult, error) {
	var result InsightResult
	if err := json.Unmarshal([]byte(response), &result); err != nil {
		return &InsightResult{
			Error: fmt.Sprintf("Failed to parse AI response: %v", err),
		}, nil
	}
	return &result, nil
}

func (s *AIService) extractTextFromContent(content models.JSONB) string {
	var textParts []string
	
	if contentArray, ok := content["content"].([]interface{}); ok {
		for _, item := range contentArray {
			if itemMap, ok := item.(map[string]interface{}); ok {
				text := s.extractTextFromNode(itemMap)
				if text != "" {
					textParts = append(textParts, text)
				}
			}
		}
	}
	
	return strings.Join(textParts, "\n")
}

func (s *AIService) extractTextFromNode(node map[string]interface{}) string {
	var textParts []string
	
	// If this is a text node, return its text
	if nodeType, ok := node["type"].(string); ok && nodeType == "text" {
		if text, ok := node["text"].(string); ok {
			return text
		}
	}
	
	// If this node has content, recursively extract text
	if content, ok := node["content"].([]interface{}); ok {
		for _, item := range content {
			if itemMap, ok := item.(map[string]interface{}); ok {
				text := s.extractTextFromNode(itemMap)
				if text != "" {
					textParts = append(textParts, text)
				}
			}
		}
	}
	
	return strings.Join(textParts, " ")
}

func (s *AIService) gatherUserData(userID uuid.UUID, limit int) (map[string]interface{}, error) {
	data := make(map[string]interface{})
	
	// Get recent notes
	var notes []models.Note
	if err := s.db.Where("user_id = ?", userID).
		Order("updated_at DESC").
		Limit(limit).
		Find(&notes).Error; err != nil {
		return nil, fmt.Errorf("failed to fetch notes: %w", err)
	}
	
	// Get people
	var people []models.Person
	if err := s.db.Where("user_id = ?", userID).Find(&people).Error; err != nil {
		return nil, fmt.Errorf("failed to fetch people: %w", err)
	}
	
	// Get todos
	var todos []models.Todo
	if err := s.db.Joins("JOIN notes ON todos.note_id = notes.id").
		Where("notes.user_id = ?", userID).
		Order("todos.created_at DESC").
		Limit(limit).
		Find(&todos).Error; err != nil {
		return nil, fmt.Errorf("failed to fetch todos: %w", err)
	}
	
	// Get connections
	var connections []models.Connection
	if err := s.db.Where("user_id = ?", userID).
		Order("created_at DESC").
		Limit(limit).
		Find(&connections).Error; err != nil {
		return nil, fmt.Errorf("failed to fetch connections: %w", err)
	}
	
	data["notes"] = notes
	data["people"] = people
	data["todos"] = todos
	data["connections"] = connections
	data["summary"] = map[string]int{
		"total_notes":       len(notes),
		"total_people":      len(people),
		"total_todos":       len(todos),
		"total_connections": len(connections),
	}
	
	return data, nil
}

File: desktop-client/src/services/aiService.ts
import axios, { AxiosInstance } from 'axios';

export interface AIProvider {
  id: string;
  name: string;
  baseUrl: string;
  requiresApiKey: boolean;
}

export interface AIConfig {
  provider: string;
  apiKey: string;
  enabled: boolean;
}

export interface TodoExtractionResult {
  todos: Array<{
    text: string;
    assignedPerson?: string;
    dueDate?: string;
    priority?: 'low' | 'medium' | 'high';
  }>;
  confidence: number;
}

export interface PeopleMentionAnalysis {
  mentions: Array<{
    name: string;
    context: string;
    confidence: number;
    suggestedPersonId?: string;
  }>;
  relationships: Array<{
    person1: string;
    person2: string;
    relationshipType: string;
    confidence: number;
  }>;
}

export interface InsightResult {
  insights: Array<{
    type: 'pattern' | 'suggestion' | 'connection' | 'trend';
    title: string;
    description: string;
    confidence: number;
    actionable?: boolean;
    relatedNotes?: string[];
    relatedPeople?: string[];
  }>;
  summary: string;
}

export const AI_PROVIDERS: AIProvider[] = [
  {
    id: 'openai',
    name: 'OpenAI GPT',
    baseUrl: 'https://api.openai.com/v1',
    requiresApiKey: true,
  },
  {
    id: 'gemini',
    name: 'Google Gemini',
    baseUrl: 'https://generativelanguage.googleapis.com/v1',
    requiresApiKey: true,
  },
  {
    id: 'grok',
    name: 'Grok (X.AI)',
    baseUrl: 'https://api.x.ai/v1',
    requiresApiKey: true,
  },
];

class AIService {
  private client: AxiosInstance | null = null;
  private config: AIConfig | null = null;
  private isAvailable: boolean = false;

  constructor() {
    // Don't auto-load config in constructor to avoid issues in tests
    // Config will be loaded when needed
  }

  async initialize(): Promise<void> {
    await this.loadConfig();
  }

  private async loadConfig(): Promise<void> {
    try {
      // Load AI config from secure storage via IPC
      const config = await window.electronAPI?.getAIConfig();
      if (config) {
        this.config = config;
        this.initializeClient();
      }
    } catch (error) {
      console.warn('Failed to load AI config:', error);
      this.isAvailable = false;
    }
  }

  private initializeClient(): void {
    if (!this.config || !this.config.enabled) {
      this.isAvailable = false;
      return;
    }

    const provider = AI_PROVIDERS.find(p => p.id === this.config!.provider);
    if (!provider) {
      console.error('Unknown AI provider:', this.config.provider);
      this.isAvailable = false;
      return;
    }

    this.client = axios.create({
      baseURL: provider.baseUrl,
      headers: {
        'Authorization': `Bearer ${this.config.apiKey}`,
        'Content-Type': 'application/json',
      },
      timeout: 30000, // 30 second timeout
    });

    this.isAvailable = true;
  }

  async updateConfig(config: AIConfig): Promise<void> {
    try {
      // Save config securely via IPC
      await window.electronAPI?.setAIConfig(config);
      this.config = config;
      this.initializeClient();
    } catch (error) {
      console.error('Failed to update AI config:', error);
      throw new Error('Failed to save AI configuration');
    }
  }

  getConfig(): AIConfig | null {
    return this.config;
  }

  isServiceAvailable(): boolean {
    return this.isAvailable && this.config?.enabled === true;
  }

  async testConnection(): Promise<boolean> {
    if (!this.isServiceAvailable() || !this.client) {
      return false;
    }

    try {
      const provider = AI_PROVIDERS.find(p => p.id === this.config!.provider);
      
      if (provider?.id === 'openai') {
        await this.client.get('/models');
      } else if (provider?.id === 'gemini') {
        await this.client.get('/models');
      } else if (provider?.id === 'grok') {
        await this.client.get('/models');
      }
      
      return true;
    } catch (error) {
      console.error('AI service connection test failed:', error);
      return false;
    }
  }

  async extractTodos(noteContent: string): Promise<TodoExtractionResult> {
    if (!this.isServiceAvailable()) {
      throw new Error('AI service is not available');
    }

    try {
      const prompt = this.buildTodoExtractionPrompt(noteContent);
      const response = await this.makeAIRequest(prompt);
      
      return this.parseTodoExtractionResponse(response);
    } catch (error) {
      console.error('Todo extraction failed:', error);
      if (!this.isServiceAvailable()) {
        throw new Error('AI service is not available');
      }
      throw new Error('Failed to extract todos from note content');
    }
  }

  async analyzePeopleMentions(noteContent: string, existingPeople: Array<{id: string, name: string}>): Promise<PeopleMentionAnalysis> {
    if (!this.isServiceAvailable()) {
      throw new Error('AI service is not available');
    }

    try {
      const prompt = this.buildPeopleMentionPrompt(noteContent, existingPeople);
      const response = await this.makeAIRequest(prompt);
      
      return this.parsePeopleMentionResponse(response);
    } catch (error) {
      console.error('People mention analysis failed:', error);
      if (!this.isServiceAvailable()) {
        throw new Error('AI service is not available');
      }
      throw new Error('Failed to analyze people mentions');
    }
  }

  async generateInsights(notes: Array<{id: string, title: string, content: string, createdAt: string}>, people: Array<{id: string, name: string}>): Promise<InsightResult> {
    if (!this.isServiceAvailable()) {
      throw new Error('AI service is not available');
    }

    try {
      const prompt = this.buildInsightGenerationPrompt(notes, people);
      const response = await this.makeAIRequest(prompt);
      
      return this.parseInsightResponse(response);
    } catch (error) {
      console.error('Insight generation failed:', error);
      if (!this.isServiceAvailable()) {
        throw new Error('AI service is not available');
      }
      throw new Error('Failed to generate insights');
    }
  }

  private async makeAIRequest(prompt: string): Promise<string> {
    if (!this.client || !this.config) {
      throw new Error('AI client not initialized');
    }

    const provider = AI_PROVIDERS.find(p => p.id === this.config!.provider);
    
    if (provider?.id === 'openai') {
      const response = await this.client.post('/chat/completions', {
        model: 'gpt-3.5-turbo',
        messages: [{ role: 'user', content: prompt }],
        temperature: 0.3,
        max_tokens: 1000,
      });
      return response.data.choices[0].message.content;
    } else if (provider?.id === 'gemini') {
      const response = await this.client.post('/models/gemini-pro:generateContent', {
        contents: [{ parts: [{ text: prompt }] }],
      });
      return response.data.candidates[0].content.parts[0].text;
    } else if (provider?.id === 'grok') {
      const response = await this.client.post('/chat/completions', {
        model: 'grok-beta',
        messages: [{ role: 'user', content: prompt }],
        temperature: 0.3,
        max_tokens: 1000,
      });
      return response.data.choices[0].message.content;
    }

    throw new Error('Unsupported AI provider');
  }

  private buildTodoExtractionPrompt(noteContent: string): string {
    return `
Analyze the following note content and extract actionable todo items. Return the result as JSON.

Note content:
${noteContent}

Please identify:
1. Clear action items or tasks
2. Assigned people (if mentioned with @)
3. Due dates (if mentioned)
4. Priority level (if indicated)

Return JSON format:
{
  "todos": [
    {
      "text": "Complete the report",
      "assignedPerson": "John Smith",
      "dueDate": "2024-01-15",
      "priority": "high"
    }
  ],
  "confidence": 0.85
}

Only extract clear, actionable items. Ignore vague statements or completed tasks.
`;
  }

  private buildPeopleMentionPrompt(noteContent: string, existingPeople: Array<{id: string, name: string}>): string {
    const peopleList = existingPeople.map(p => `- ${p.name} (ID: ${p.id})`).join('\n');
    
    return `
Analyze the following note content for people mentions and relationships. Return the result as JSON.

Note content:
${noteContent}

Existing people in database:
${peopleList}

Please identify:
1. People mentioned in the text (including @mentions)
2. Relationships between people
3. Match mentions to existing people when possible

Return JSON format:
{
  "mentions": [
    {
      "name": "John Smith",
      "context": "discussed project timeline",
      "confidence": 0.9,
      "suggestedPersonId": "existing-person-id-if-match"
    }
  ],
  "relationships": [
    {
      "person1": "John Smith",
      "person2": "Sarah Johnson",
      "relationshipType": "colleague",
      "confidence": 0.8
    }
  ]
}
`;
  }

  private buildInsightGenerationPrompt(notes: Array<{id: string, title: string, content: string, createdAt: string}>, people: Array<{id: string, name: string}>): string {
    const noteSummary = notes.slice(0, 10).map(n => `- ${n.title} (${n.createdAt})`).join('\n');
    const peopleList = people.slice(0, 20).map(p => p.name).join(', ');
    
    return `
Analyze the user's knowledge base and generate insights. Return the result as JSON.

Recent notes:
${noteSummary}

People in network:
${peopleList}

Please identify:
1. Patterns in note-taking or topics
2. Suggestions for better organization
3. Important connections between notes/people
4. Trends over time

Return JSON format:
{
  "insights": [
    {
      "type": "pattern",
      "title": "Frequent Meeting Notes",
      "description": "You create many meeting notes but rarely follow up on action items",
      "confidence": 0.8,
      "actionable": true,
      "relatedNotes": ["note-id-1", "note-id-2"]
    }
  ],
  "summary": "Your knowledge base shows strong focus on project management with opportunities for better task tracking."
}

Focus on actionable insights that help improve productivity and knowledge management.
`;
  }

  private parseTodoExtractionResponse(response: string): TodoExtractionResult {
    try {
      const parsed = JSON.parse(response);
      return {
        todos: parsed.todos || [],
        confidence: parsed.confidence || 0.5,
      };
    } catch (error) {
      console.error('Failed to parse todo extraction response:', error);
      return { todos: [], confidence: 0 };
    }
  }

  private parsePeopleMentionResponse(response: string): PeopleMentionAnalysis {
    try {
      const parsed = JSON.parse(response);
      return {
        mentions: parsed.mentions || [],
        relationships: parsed.relationships || [],
      };
    } catch (error) {
      console.error('Failed to parse people mention response:', error);
      return { mentions: [], relationships: [] };
    }
  }

  private parseInsightResponse(response: string): InsightResult {
    try {
      const parsed = JSON.parse(response);
      return {
        insights: parsed.insights || [],
        summary: parsed.summary || 'No insights available.',
      };
    } catch (error) {
      console.error('Failed to parse insight response:', error);
      return { insights: [], summary: 'Failed to generate insights.' };
    }
  }
}

export const aiService = new AIService();